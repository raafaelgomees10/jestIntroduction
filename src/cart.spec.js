import Cart from './cart';

describe('Cart', () => {
  let cart;
  let product = {
    title: 'Tenis Ous',
    price: 35388,
  };

  let product2 = {
    title: 'Shape Element',
    price: 41872,
  };

  beforeEach(() => {
    cart = new Cart();
  });

  describe('getTotal()', () => {
    it('Should return 0 when getTotal is executed in a newly created instance', () => {
      expect(cart.getTotal().getAmount()).toEqual(0);
    });

    it('Should multiply quantity and price and receive the total amount', () => {
      const item = {
        product,
        quantity: 2,
      };
      cart.add(item);
      expect(cart.getTotal().getAmount()).toEqual(70776);
    });

    it('Should ensure no more than on product exists at a time', () => {
      cart.add({
        product,
        quantity: 2,
      });

      cart.add({
        product,
        quantity: 1,
      });
      expect(cart.getTotal().getAmount()).toEqual(35388);
    });

    it('Should update total when a product gets included and then removed', () => {
      cart.add({
        product,
        quantity: 2,
      });

      cart.add({
        product: product2,
        quantity: 1,
      });

      cart.remove(product);

      expect(cart.getTotal().getAmount()).toEqual(41872);
    });
  });

  describe('checkout()', () => {
    it('Should return an object with the total and the list of items', () => {
      cart.add({
        product,
        quantity: 2,
      });

      cart.add({
        product: product2,
        quantity: 3,
      });

      expect(cart.checkout()).toMatchSnapshot();
    });

    it('Should return an object with the total and the list of items when sumary() is called', () => {
      cart.add({
        product,
        quantity: 2,
      });

      cart.add({
        product: product2,
        quantity: 3,
      });

      expect(cart.sumary()).toMatchSnapshot();
      expect(cart.getTotal().getAmount()).toBeGreaterThan(0);
    });

    it('Should reset the cart when checkout() is called', () => {
      cart.add({
        product: product2,
        quantity: 3,
      });
      cart.checkout();
      expect(cart.getTotal().getAmount()).toEqual(0);
    });
  });
});
