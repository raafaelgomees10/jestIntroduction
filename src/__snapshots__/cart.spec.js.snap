// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart checkout Should return an object with the total and the list of items 1`] = `
{
  "items": [
    {
      "product": {
        "price": 35388,
        "title": "Tenis Ous",
      },
      "quantity": 2,
    },
    {
      "product": {
        "price": 41872,
        "title": "Shape Element",
      },
      "quantity": 3,
    },
  ],
  "total": 196392,
}
`;

exports[`Cart checkout Should return an object with the total and the list of items when sumary() is called 1`] = `
{
  "items": [
    {
      "product": {
        "price": 35388,
        "title": "Tenis Ous",
      },
      "quantity": 2,
    },
    {
      "product": {
        "price": 41872,
        "title": "Shape Element",
      },
      "quantity": 3,
    },
  ],
  "total": 196392,
}
`;
